<?xml version="1.0" ?>
<launch>
	<!-- If mode = sim: Simulated AUV with MBES and control in loaded bathymetric map 
	If mode = gt: reproduce Hugin survey and construct map of real bathymetry --> 
    <arg name="mode" default="sim"/>
	<arg name="namespace" default="hugin"/>
	<arg name="debug" default="0"/>

	<!-- Rates for simulating missions  -->
    <arg name="odom_rate" default="0.1"/>
	<arg name="meas_rate" default="0.5"/>
    <arg name="replay_rate" default="0.1"/>
	
	<!-- MBES sensor -->
	<arg name="n_beams_mbes" default="40"/>	
	<arg name="mbes_open_angle" value="1.308" />
	<!-- For lost targets -->
	<!-- <arg name="mbes_open_angle" value="1.308" /> -->

	<!-- Car detection parameters -->
    <arg name="detection" default="False"/>
    <arg name="add_mini" default="False"/>
    
    <!-- Particle filter -->
    <arg name="particle_filter" default="True"/>
    <arg name="filter" default="pf"/>

    <arg name="map_name" default="demo_02"/> <!-- Name of UFO map to be loaded (if exists) or saved when gt mission is over -->

    <!-- Real data  -->
    <arg name="path" default="/media/torroba/Seagate Expansion Drive/MMTDatasets/LostTargets/Ping_Processed/5-Products/MBES_GSF/Post_Deployment_PROCESSED"/>
    <arg name="cereal_trajectory" default="$(arg path)/mbes_pings.cereal"/>

	<!-- Coordinates of odom frame wrt the map frame. By default, the AUV start the mission here -->
	<arg name="x" default="-200.0"/>
	<arg name="y" default="-20.0"/>
	<arg name="z" default="0.0"/>
	<arg name="roll" default="0.0"/>
	<arg name="pitch" default="0.0"/>
	<arg name="yaw" default="3.0"/>

	<group ns="$(arg namespace)">

		<!-- Particle filter -->
		<group if="$(arg particle_filter)">
			<include file="$(find auv_particle_filter)/launch/auv_pf.launch" >
				<arg name="mode" value="$(arg mode)"/>
				<arg name="filter" value="$(arg filter)"/>
				<arg name="namespace" value="$(arg namespace)"/>
				<arg name="n_beams_mbes" value="$(arg n_beams_mbes)"/>
				<arg name="mbes_open_angle" value="$(arg mbes_open_angle)" />
			</include>
		</group>

		<!-- Bathymetry change detection -->
		<group if="$(arg detection)">
			<include file="$(find change_detection)/launch/change_detection.launch" >
				<arg name="mode" value="$(arg mode)"/>
				<arg name="namespace" value="$(arg namespace)"/>
				<arg name="filter" value="$(arg filter)"/>
			</include>
		</group>

		<!-- Hugin model -->
		<param name="robot_description" command="$(find xacro)/xacro '$(find hugin_description)/robots/hugin_auv_default.urdf.xacro' debug:=$(arg debug) namespace:=$(arg namespace)" />
		<node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="true" output="screen">
			<!-- <param name="robot_description" value="/$(arg namespace)/robot_description" /> -->
		</node>

		<!-- Publish map as pcl  -->
		<node type="map_pcl_vis.py" pkg="map_pcl_publisher" name="map_pub" output="screen">
			<param name="map_cloud_path"  value="/home/torroba/Downloads/post_deployment/KTH_Post_Deployment_AVG_WGS84UTM32N_RH200_50cm.xyz" />
			<param name="map_gp_path"  value="$(find auv_particle_filter)/meshes/posterior.npy"/>
			<param name="pings_path"  value="/home/torroba/Downloads/post_deployment/mbes_pings.cereal" />
			<param name="map_frame" value="map" />
		</node>

		<!-- Simulate Hugin AUV -->
		<group if="$(eval mode == 'sim')">
    		<node pkg="tf" type="static_transform_publisher" name="tf_world_map_pub" args="0 0 0 0 0 0 world map 50" />

    		<node pkg="tf" type="static_transform_publisher" name="tf_map_odom_pub" args="$(arg x) $(arg y) $(arg z) $(arg yaw) $(arg pitch) $(arg roll) map odom 50" />

			<node pkg="auv_model" type="auv_motion_model" name="auv_motion_model" output="screen">
				<param name="odom_rate" value="$(arg odom_rate)" />
				<param name="meas_rate" value="$(arg meas_rate)" />
				<param name="odom_sim" value="/$(arg mode)/odom" />
				<param name="world_frame" value="world" />
				<param name="map_frame" value="map" />
				<param name="odom_frame" value="odom" />
				<param name="base_link" value="$(arg namespace)/base_link" />
				<param name="mbes_link" value="$(arg namespace)/mbes_link" />
				<param name="throttle_cmd" value="/$(arg namespace)/throttle_cmd" />
				<param name="thruster_cmd" value="/$(arg namespace)/thruster_cmd" />
				<param name="inclination_cmd" value="/$(arg namespace)/inclination_cmd" />
				<param name="mbes_sim_as" value="/mbes_sim_server" />
				<param name="n_beams_mbes" value="$(arg n_beams_mbes)" />
			</node>

			<node pkg="auv_model" type="auv_sim_teleop.py" name="auv_sim_teleop" output="screen">
				<param name="throttle_cmd" value="/$(arg namespace)/throttle_cmd" />
				<param name="thruster_cmd" value="/$(arg namespace)/thruster_cmd" />
				<param name="inclination_cmd" value="/$(arg namespace)/inclination_cmd" />
			</node>
  		</group>

		<!-- Reproduce real Hugin survey -->
  		<group unless="$(eval mode == 'sim')">
			<node pkg="auv_2_ros" type="auv_2_ros" name="auv_2_ros" output="screen">
				<param name="trajectory_cereal" value="$(arg cereal_trajectory)"/>
				<param name="replay_rate" value="$(arg replay_rate)" />
				<param name="mbes_pings" value="/gt/mbes_pings" />
				<param name="map_pcl" value="/gt/map" />
				<param name="debug_pings" value="debug/mbes_pings" />
				<param name="mbes_pings_topic" value="/gt/mbes_pings" />
				<param name="world_frame" value="world" />
				<param name="map_frame" value="map" />
				<param name="odom_frame" value="odom" />
				<param name="base_link" value="$(arg namespace)/base_link" />
				<param name="mbes_link" value="$(arg namespace)/mbes_link" />
				<param name="mbes_sim_as" value="/mbes_sim_server" />
				<param name="add_mini" value="$(arg add_mini)" />
				<param name="n_beams_mbes" value="$(arg n_beams_mbes)" />
				<param name="start_mission_ping_num" value="51000" />
				<param name="end_mission_ping_num" value="72000" />
			</node>
		</group>

	</group>

</launch>
